restli-client/src/main/java/com/linkedin/restli/client/RestClient.java

```diff
@@ -414,53 +440,59 @@ public void onSuccess(ProtocolVersion protocolVersion)
   {
     TimingContextUtil.beginTiming(requestContext, FrameworkTimingKeys.CLIENT_REQUEST_RESTLI_GET_PROTOCOL.key());
     ProtocolVersionOption versionOption = request.getRequestOptions().getProtocolVersionOption();
-    ProtocolVersion announcedProtocolVersion = null;
-    // fetch server announced version only for 'ProtocolVersionOption.USE_LATEST_IF_AVAILABLE'
-    if (versionOption == ProtocolVersionOption.USE_LATEST_IF_AVAILABLE) {
-      final String serviceName = request.getServiceName();
-      // check cache first.
-      announcedProtocolVersion = _announcedProtocolVersionCache.getIfPresent(serviceName);
-      if (announcedProtocolVersion == null) {
-        // if announcedProtocolVersion is not available in cache find and cache it.
-        try {
-          _client.getMetadata(new URI(_uriPrefix + serviceName), Callbacks.handle(metadata -> {
-            ProtocolVersion announcedVersion = getAnnouncedVersion(metadata);
-            _announcedProtocolVersionCache.put(serviceName, announcedVersion);
-            final ProtocolVersion protocolVersion = getProtocolVersion(AllProtocolVersions.BASELINE_PROTOCOL_VERSION,
-                AllProtocolVersions.PREVIOUS_PROTOCOL_VERSION,
-                AllProtocolVersions.LATEST_PROTOCOL_VERSION,
-                AllProtocolVersions.NEXT_PROTOCOL_VERSION,
-                announcedVersion,
-                versionOption,
-                _forceUseNextVersionOverride);
-            TimingContextUtil.endTiming(requestContext, FrameworkTimingKeys.CLIENT_REQUEST_RESTLI_GET_PROTOCOL.key());
-            callback.onSuccess(protocolVersion);
-          }, callback));
-        } catch (URISyntaxException e) {
-          throw new RuntimeException("Failed to create a valid URI to fetch properties for!");
-        }
-        return;
-      }
+
+    // First check cache.
+    final String serviceName = request.getServiceName();
+    Pair<ProtocolVersion, Set<String>> cachedMetadata = _serviceMetadataCache.getIfPresent(serviceName);
+    if (cachedMetadata != null)
+    {
+      callback.onSuccess(Pair.of(getProtocolVersion(AllProtocolVersions.BASELINE_PROTOCOL_VERSION,
+          AllProtocolVersions.PREVIOUS_PROTOCOL_VERSION,
+          AllProtocolVersions.LATEST_PROTOCOL_VERSION,
+          AllProtocolVersions.NEXT_PROTOCOL_VERSION,
+          cachedMetadata.getLeft(),
+          versionOption,
+          _forceUseNextVersionOverride), cachedMetadata.getRight()));
+      return;
+    }
+
+    try
+    {
+      _client.getMetadata(new URI(_uriPrefix + serviceName), Callbacks.handle(metadata ->
+      {
+        // Use the metadata from the server to extract protocol version only if versionOption is USE_LATEST_IF_AVAILABLE
+        final ProtocolVersion announcedVersion = (versionOption ==
+            ProtocolVersionOption.USE_LATEST_IF_AVAILABLE) ? getAnnouncedVersion(metadata) : null;
+        final Set<String> experimentalContentTypeHeaders = getExperimentalContentTypeHeaders(metadata);
```

Comment:
Instead of call this experimental content type, can we just say these are server announced content types. Server can list them in the order of preference.

The client can then pick the first one it supports.